pkg/virt-controller/watch/migration.go

```diff
@@ -4,129 +4,258 @@ import (
 
 	dispatch := MigrationDispatch{
-		vmService: vmService,
+		restClient: restClient,
+		vmService:  vmService,
+		clientset:  clientset,
 	}
 	return &dispatch
 }
 
 type MigrationDispatch struct {
-	vmService services.VMService
+	restClient *rest.RESTClient
+	vmService  services.VMService
+	clientset  *kubernetes.Clientset
 }
 
 func (md *MigrationDispatch) Execute(store cache.Store, queue workqueue.RateLimitingInterface, key interface{}) {
 
-	setMigrationFailed := func(migrationCopy *v1.Migration) {
+	setMigrationPhase := func(migration *v1.Migration, phase v1.MigrationPhase) error {
+
+		if migration.Status.Phase == phase {
+			return nil
+		}
 
-		logger := logging.DefaultLogger().Object(migrationCopy)
-		migrationCopy.Status.Phase = v1.MigrationFailed
+		logger := logging.DefaultLogger().Object(migration)
+
+		// Copy migration for future modifications
+		migrationCopy, err := copy(migration)
+		if err != nil {
+			logger.Error().Reason(err).Msg("could not copy migration object")
+			queue.AddRateLimited(key)
+			return nil
+		}
+
+		migrationCopy.Status.Phase = phase
 		// TODO indicate why it was set to failed
-		err := md.vmService.UpdateMigration(migrationCopy)
+		err = md.vmService.UpdateMigration(migrationCopy)
 		if err != nil {
 			logger.Error().Reason(err).Msgf("updating migration state failed: %v ", err)
 			queue.AddRateLimited(key)
-			return
+			return err
 		}
 		queue.Forget(key)
+		return nil
+	}
+
+	setMigrationFailed := func(mig *v1.Migration) {
+		setMigrationPhase(mig, v1.MigrationFailed)
 	}
 
 	obj, exists, err := store.GetByKey(key.(string))
 	if err != nil {
 		queue.AddRateLimited(key)
 		return
 	}
-	if exists {
-		var migration *v1.Migration = obj.(*v1.Migration)
-		logger := logging.DefaultLogger().Object(migration)
+	if !exists {
+		queue.Forget(key)
+		return
+	}
 
-		if migration.Status.Phase == v1.MigrationUnknown {
-			// Copy migration for future modifications
-			migrationCopy, err := copy(migration)
-			if err != nil {
-				logger.Error().Reason(err).Msg("could not copy migration object")
-				queue.AddRateLimited(key)
-				return
-			}
-			// Fetch vm which we want to migrate
-			vm, exists, err := md.vmService.FetchVM(migration.Spec.Selector.Name)
-			if err != nil {
-				logger.Error().Reason(err).Msgf("fetching the vm %s failed", migration.Spec.Selector.Name)
-				queue.AddRateLimited(key)
+	var migration *v1.Migration = obj.(*v1.Migration)
+	logger := logging.DefaultLogger().Object(migration)
+
+	vm, exists, err := md.vmService.FetchVM(migration.Spec.Selector.Name)
+	if err != nil {
+		logger.Error().Reason(err).Msgf("fetching the vm %s failed", migration.Spec.Selector.Name)
+		queue.AddRateLimited(key)
+		return
+	}
+
+	if !exists {
+		logger.Info().Msgf("VM with name %s does not exist, marking migration as failed", migration.Spec.Selector.Name)
```

Comment:
Did you mean to use logger.Info() or should this condition also be a call to logger.Error()